### PR TITLE
[None][perf] Prewarm DeepGEMM paged_mqa_logits_metadata JIT buckets

### PR DESCRIPTION
**Pull request by chenfeiz0326**
## Description

DSA's `Indexer.prepare_scheduler_metadata` (`tensorrt_llm/_torch/attention_backend/sparse/dsa.py`) calls `deep_gemm.get_paged_mqa_logits_metadata(context_lens, block_kv, num_sms)` on every iteration inside `_prepare_inputs`. The underlying kernel `deep_gemm::sched::smxx_paged_mqa_logits_metadata` is templated on `kAlignedBatchSize = ceil(num_generations, 32)`, and deep_gemm's Python-side JIT (`deep_gemm_cpp_tllm.so`) compiles a fresh cubin — spawning `nvcc` → `cicc` → `ptxas` — the first time each bucket is requested (~3s per bucket on Blackwell).

CUDA-graph warmup exercises only the batch sizes in `cuda_graph_batch_sizes`. On a typical `max_batch_size=512` config, those round up to only `{32, 64, 96, 128, 192, 256, 320, 384, 448, 512}` — leaving `{160, 224, 288, 352, 416, 480}` uncompiled. Any inference iter whose `num_generations` lands on an uncovered bucket eats a ~3s stall inside `_prepare_inputs` on first touch, and the stall repeats every time a fresh container process starts (the JIT cache is either process-local `LruCache` or an on-disk cache at `$DG_JIT_CACHE_DIR`, which defaults to a container-ephemeral path).

### Observed behavior (before fix)

**Case A — DeepSeek V3.2 FP4 `ep:4-gpus:4` on lyris GB300 (SM103 aarch64, num_sms=152):**

| rep | total_token_throughput (tok/s) |
|---:|---:|
| 1 (fresh container) | 6,494 |
| 2 (same container) | 10,730 |
| 3 (same container) | 10,727 |

**Case B — DeepSeek V3.2 FP4 `ep:8-gpus:8` on bia B300 (SM103 x86_64, num_sms=148):**

| rep | total_token_throughput (tok/s) |
|---:|---:|
| 1 (fresh container) | 4,233 |
| 2 (same container) | 9,761 |
| 3 (same container) | 9,749 |

nsys of Case A rep 1 iters 140-150 (`TLLM_PROFILE_START_STOP=140-150`):

| iter | `_prepare_inputs` | `_forward_step` |
|---:|---:|---:|
| 140 | **3,092 ms** | 388 ms |
| 141 | 22 ms | 501 ms |
| 142 | 21 ms | 416 ms |
| 143 | 22 ms | 430 ms |
| 144 | **3,144 ms** | 422 ms |
| 145 | 21 ms | 392 ms |
| 149 | **3,158 ms** | 416 ms |

`num_generations` at slow iters: 139 (→ aligned 160), 198 (→ 224), 268 (→ 288) — exactly the buckets warmup missed. CPU sampling during the stalls is dominated by `nvvm/bin/cicc` (14,932 samples across the 3 slow iters), `cc1plus`, `nvcc`, `ptxas` — the deep_gemm Python-side JIT firing on first touch. Warm autotuner / TorchInductor / CuTe DSL persistent caches all had no effect (verified by bisection); only `$DG_JIT_CACHE_DIR` populated the missing cubins.

### Fix

After `_run_cuda_graph_warmup`, call `get_paged_mqa_logits_metadata` for every 32-aligned batch bucket up to `max_batch_size` using a zero-filled dummy `context_lens` tensor. This funnels the missing compiles into the deterministic warmup phase.

- **Scope**: `_torch/pyexecutor/model_engine.py` — one new method (`_warmup_dg_paged_mqa_logits_metadata`) plus one call site after CUDA-graph warmup. Guarded on `isinstance(self.attn_metadata, DSAtrtllmAttentionMetadata)`, so it is a strict no-op on non-DSA models.
- **Cost**: one `nvcc` invocation per bucket the CUDA-graph warmup didn't already trigger (typically 5-6 extra compiles for `max_batch_size=512`), added to warmup wall-clock (~15-30s). No runtime cost.
- **Benefit**: eliminates the per-fresh-process 3s stalls at first-touch iters.

### Test Coverage

Validated on two DeepSeek V3.2 FP4 workloads across two different clusters and two different topologies:

**Case A — lyris GB300, `ep:4-gpus:4`** (empty `$DG_JIT_CACHE_DIR`, so rep 1 cannot cheat via disk):

| rep | total_token_throughput (tok/s) |
|---:|---:|
| 1 (patched) | **10,516** |
| 2 (patched) | 10,360 |

Rep2/Rep1: 0.99× (down from 1.65× / +62% rep-1 recovery).

**Case B — bia B300, `ep:8-gpus:8`** (fresh container, wheel reuse across baseline/patched):

| rep | total_token_throughput (tok/s) |
|---:|---:|
| 1 (patched) | **9,793** |
| 2 (patched) | 9,743 |
| 3 (patched) | 9,786 |

Rep2/Rep1: 0.99× (down from 2.31× / +131% rep-1 recovery). CV across 3 reps: 0.24% (down from ~40%).

Both runs show identical warmup evidence in bench logs:
```
[TRT-LLM] [I] [_torch][RANK 0] [DG warmup] Pre-compiling paged_mqa_logits_metadata for 16 aligned batch buckets up to 512 (block_kv=64, num_sms={148,152})
```

The existing `perf-sanity` DSA tests (e.g. `deepseek_v3.2_fp4-...ep:{4,8}-gpus:{4,8}`) exercise this code path; the fix should show up as improved / more stable first-rep throughput on any DSA benchmark that reports rep-1 numbers.

## PR Checklist

- [x] PR description clearly explains what and why.
- [x] PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of my knowledge.
- [x] No API changes.
- [x] No new dependencies.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1092,6 +1101,72 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
+        except ImportError:
+            logger.info(
+                "[DG warmup] deep_gemm.get_paged_mqa_logits_metadata not "
+                "available; skipping paged_mqa_logits_metadata prewarm.")
+            return
+
+        num_sms = attn_meta.num_sms
+        max_bs = max(1, int(self.batch_size))
+        max_aligned = ((max_bs + 31) // 32) * 32
+        buckets = list(range(32, max_aligned + 32, 32))
```
Comment (hyukn) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1176]: Can any shape outside the buckets still trigger the JIT issue?

Comment (chenfeiz0326) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1176]: Yes. The previous prewarm only covered buckets up to `max_batch_size`, but `context_lens.size(0)` can exceed that on two paths:

- `use_expanded_buffers_for_mtp=True` (MTP): `dsa.py` L842 calls the kernel with `num_generations * (1 + max_draft_tokens)` — up to `4 × max_batch_size` for `max_draft_tokens=3`.
- `expand_for_dsl=True`: `dsa.py` L854 calls it with `num_generations * dsl_expand_factor`.

The JIT template key is `<align(context_lens.size(0), 32), split_kv, num_sms>` (from `smxx_paged_mqa_logits_metadata.hpp` — `split_kv = block_kv * 4` is fixed at 256, `num_sms` is device-fixed), so those uncovered buckets would each pay the ~3s first-touch stall.

Fixed in d5bdb9548d — bucket ceiling is now `max_batch_size × max(1, 1 + max_draft_tokens if MTP-expanded, dsl_expand_factor if DSL-expanded)`. No-op for the plain DeepSeek V3.2 FP4 configs already covered by the perf numbers in the PR description.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -1092,6 +1103,92 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
+        for aligned_bs in buckets:
+            # Kernel scans `context_lens` and prefix-sums schedules; a
+            # zero-filled 2D tensor of shape (aligned_bs, 1) is enough to
+            # trigger dispatch and compile — the metadata output is
+            # discarded.
+            dummy = torch.zeros(aligned_bs, 1, dtype=torch.int32, device="cuda")
+            try:
+                _ = get_paged_mqa_logits_metadata(dummy, _DG_SCHEDULE_BLOCK_KV,
+                                                  num_sms)
+            except Exception as e:
```
Comment (tongyuantongyu) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Suggest narrow this to `except RuntimeError`, so unexpected exceptions, like `ValueError` caused by upstream changing the signature of `get_paged_mqa_logits_metadata` won't be swallowed here.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -1092,6 +1103,92 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
+        # metadata; default to 1 when the attribute is missing / disabled.
+        expand_factor = 1
+        if getattr(attn_meta, "use_expanded_buffers_for_mtp", False):
+            expand_factor = max(
+                expand_factor,
+                1 + int(getattr(attn_meta, "max_draft_tokens", 0) or 0))
+        if getattr(attn_meta, "expand_for_dsl", False):
+            expand_factor = max(
+                expand_factor,
+                int(getattr(attn_meta, "dsl_expand_factor", 1) or 1))
```
Comment (YihuiLu512) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: The value of `dsl_expand_factor` is affected by the previous iteration. This is not an issue when MTP is disabled or CUDA graph is enabled. However, consider the combination of “DSL + FP4 + MTP-3 + CUDA graph off”:

- “DSL” causes the code to enter this path.
- “FP4 + MTP-3” makes the runtime `min_factor` equal to 2.
- “CUDA graph off” makes the `factor` equal to 1 at this point, **which leads to an underestimation problem during warmup.**

Should we consider skipping this warmup when CUDA graph is disabled, to keep the behavior consistent (either no warmup at all or full warmup, rather than partial warmup — which was also the original intention of this PR)?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -1092,6 +1103,92 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
+        try:
+            from tensorrt_llm.deep_gemm import get_paged_mqa_logits_metadata
+        except ImportError:
+            logger.info(
+                "[DG warmup] deep_gemm.get_paged_mqa_logits_metadata not "
+                "available; skipping paged_mqa_logits_metadata prewarm.")
+            return
+
+        num_sms = attn_meta.num_sms
+        max_bs = max(1, int(self.batch_size))
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1163]: Beam search expands each generation request into multiple sequence rows, and the size could be `max_batch_size * max_beam_width`. Could this leave buckets above `max_batch_size` uncovered?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (chienchunhung) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: +1 to @YihuiLu512 's point. 

Is it possible to calculate a stable upper bound inside the DSA layer, or enumerate all possible outcomes rather than reading the last batch’s factor.

**Review by chienchunhung — APPROVED** [review]
The prewarm approach seems promising. Thanks for the PR!

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 2 above)
Comment (chenfeiz0326) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Good call — narrowed to `except RuntimeError` in df475f94d2 with a comment explaining why signature drifts (`TypeError` / `ValueError`) should surface loudly instead of silently degrading perf. Thanks!

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (chenfeiz0326) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Sharp catch — you're right, and I picked the "full warmup" side of your two options in df475f94d2 so the CUDA-graph-off path is covered too.

Switched to a static upper bound instead of reading the runtime `dsl_expand_factor` off `attn_metadata`. `_pick_dsl_expand` returns `factor = next_n // eff` for `eff ∈ kernel_atoms`, so `factor ≤ next_n = 1 + max_draft_tokens` at `eff == 1` — that's the worst case regardless of `min_factor`, CUDA-graph state, or which iter's pick we happen to observe. The MTP and DSL branches both collapse to the same ceiling `1 + max_draft_tokens`, so the DSL + FP4 + MTP-3 + CUDA-graph-off combo you flagged now gets buckets warmed all the way up to `max_bs × (1 + max_draft_tokens)`.

Docstring updated to explain why we don't read the live `dsl_expand_factor`.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 4 above)
Comment (chenfeiz0326) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1163]: Fair point — DSA doesn't currently thread `beam_width` through `attn_metadata` (grep for `beam_width` in `dsa.py` returns nothing), and DSA users don't run beam in practice today, so this is a no-op for the configs the PR is targeting. But the fix is one line and cheap, so in df475f94d2 the bucket ceiling is now `max_bs × max_beam_width × expand_factor` as defense-in-depth for future beam + DSA — no silent uncovered range if that combination ever ships. Thanks!

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (chenfeiz0326) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Yes — went with the stable upper bound in df475f94d2 rather than skipping the warmup when CUDA graph is off. `_pick_dsl_expand` returns `factor = next_n // eff` for `eff ∈ kernel_atoms`; the worst case is `eff == 1`, giving `factor = next_n = 1 + max_draft_tokens`. So `1 + max_draft_tokens` bounds every possible runtime pick, independent of `min_factor` / batch size / max_ctx / whichever iter's pick we happen to read. That folds the MTP and DSL branches into a single ceiling and keeps the warmup enabled uniformly. Thanks for the pointer!

**Comment by chenfeiz0326** [comment]
/bot skip --comment "No need to run the whole CI pipeline again"
